### ISSUE TITLE
follower read reliability

### ISSUE THREAD

**Issue by mittalrishabh** [tikv/tikv#17158]
## Feature Request
This ticket is to track all the issues to open to improve reliability of follower read. 
### Is your feature request related to a problem? Please describe:

There are three issues
1. https://github.com/tikv/tikv/issues/16960 : Retry storm of read index if leader is slow
2. https://github.com/tikv/tikv/issues/17058 : Dedupe read index requests to the leader
3. Check for kv read timeout timeout before retrying the requests.  It is the extension of first issue. 
### Describe the feature you'd like:

1. Prevention of retry storm
2. Dedup of read index request

**Comment by mittalrishabh** [comment]
I would like to discuss the possible solutions to tackle above problems 
Solution 1 : Reusing the result for the request with the same ts and region id if there are no memory locks on that region. It will help in scenarios with read-heavy and write-light operations. It helps in both issue 1 and issue 2. 
Solution 2 : Batch the read index request messages into one message. Number of messages sent to the leader will be reduced regardless of the write traffic. However it won't help in reducing the latency.

**Comment by mittalrishabh** [comment]
I can add a metric to see how many times we find no lock on a region for our traffic if we want to proceed with solution 1.

**Comment by cfzjywxk** [comment]
> I can add a metric to see how many times we find no lock on a region for our traffic if we want to proceed with solution 1.

Yes, through metrics and testing environment validation, we can determine the ROI of these improvements.

**Comment by mittalrishabh** [comment]
added this metric https://github.com/tikv/tikv/pull/17220
